### PR TITLE
Milestone 5: Friendships v1

### PR DESCRIPTION
**Pull request by edxco**
In this milestone we implemented the following features:

- Friendship mutual associations.
- Send Friend request feature in views.
- Confirm the Friend request feature in views.
- Add custom methods to the User model.
- Rspec tests for models and features using Capybara.

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -1,3 +1,4 @@
+# rubocop:disable all
 class User < ApplicationRecord
   # Include default devise modules. Others available are:
```
Comment (bertil291utn) [inline app/models/user.rb]: Disabled linters
- Remove statement that disables linters, because it's not good practice to disable them

**Review by bertil291utn — CHANGES_REQUESTED** [review]
# **Status:** Changes requested 🟠 
<p align="center"> image <br><strong>Not yet</strong> </p>

# Hey guys, you've been doing a great job, and also I hope you enjoyed coding.

### Your work is amazing you master it GitHub Flow you have a professional README file. 

However, in order to accomplish as the actual page, you required to do these changes, listed below:

### General requirements
Edit and change the branch name, the name should be one that defines this work
### Specific requirements 
* [ ] There is a button to invite the user to friendship on the users list page and on single-user page ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/user_can_see_invite_buttton.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
  - Add a button next to see profile button
image
  -  Requirements ask about a button not a link , change it 
image

 -  Fix the next requirement and please be sure the other requirements are working
* [ ] User can invite friends ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/send_invitation.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
* [ ] User can see friendships requests ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/pending_request.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
* [ ] User has to accept friendship request to start friendship ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/accept_reject_friendship.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
* [ ] User can reject a friendship request ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/accept_reject_friendship.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))

image


### OPTIONAL
- Add the current user name in the navbar element, to know which user is current logged in 

### Summary
- Fix the options the send a friendship request
- As is not working the friendship request function, the other requirements I can't review, please work on it to review on the next one 
-  Don't forget to work with inline comments below this comment box.

***

Guys work on the requested changes above and request another review.
Good luck. 

> “Never stop, never conform until what is good becomes even better, and what’s better becomes excellent.”

**Review by acushlakoncept — CHANGES_REQUESTED** [review]
Looking at the previous review, I can't help notice the significant improvement on your codebase, that's very commendable, well done! 👏 

I noticed you missed out on the part you were asked to "Remove statement that disables linters, because it's not good practice to disable them"
Inside your `User Model` I notice you still have `# rubocop:disable all` please kindly remove it. Rubocop is there to help us write better codes by enforcing some of its rules, if something is wrong find a way to refactor your code, you will be surprised how better you could become by aligning yourself to some of its guidelines.

- [x] I am sorry, the previous TSE forgot to mention this, but it's absolutely necessary that you maintain some of these good coding practices in order to be better as we progress along with our software development career.
* There should be absolutely NO logic in views. >Views are only for displaying information that became available to it from its corresponding controller not generating it. 
Affected files:
> `layouts/application.html.erb`
> `users/_received_request.html.erb`
> `users/_user.html.erb`
> `users/index.html.erb`
> `users/show.html.erb`

**The Resources below can come in handy**
* [Where Do I Put My Code?](http://codefol.io/posts/Where-Do-I-Put-My-Code)
* [The Beginner's Guide to Rails Helpers](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers)
* [Rails View Helpers Best Practices](https://medium.com/@acushla4real/rails-view-helpers-best-practices-guide-d7d7b6db3db4)
* [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from Pathwright
#### Kindly remove all logics from view related files and use helpers.

Great job so far guys !!! 👍 

### Cheers and Happy coding! 💻 👏

(Feel free to contact me on slack(@acushla), if you have any additional question)
Uduak Essien
[Twitter](https://twitter.com/acushlakoncept) | [LinkedIn](https://www.linkedin.com/in/acushlakoncept/)

File: app/views/layouts/application.html.erb [hunk 2]
```diff
@@ -19,21 +20,17 @@
         </div>
       </div>
       <% if current_user %>
+       Hello, <%= link_to current_user.name,  user_path(current_user), class: 'profile-link' %>!
         <%= link_to 'Sign out', destroy_user_session_path, method: :delete %>
       <% else %>
         <%= link_to 'Sign in', user_session_path %>
-      <% end %>
+      <% end %>
```
Comment (tirthajyoti-ghosh) [inline app/views/layouts/application.html.erb]: Please move this logic to helpers.

File: app/views/users/_received_request.html.erb [hunk 3]
```diff
@@ -0,0 +1,10 @@
+<h2>Received Requests</h2>
+<% @friend_requests.each do |user| %>
+<li class="my-4">
+  <% if !current_user.friend?(user) %>
+    <%= user.name %> want to be your friend.
+    <%= link_to 'Accept', user_friendship_path(@friend_requests, user), method: :put, style: "color: #008000" %>
+    <%= link_to 'Reject', user_friendship_path(@friend_requests, user), method: :delete, style: "color: #FF0000" %>
+  <% end %>
```
Comment (tirthajyoti-ghosh) [inline app/views/users/_received_request.html.erb]: Please move this logic to helpers.

File: app/views/users/_user.html.erb [hunk 4]
```diff
@@ -1,6 +1,18 @@
...
+    <% if current_user.friends.none?(user) && current_user.pending_friends.none?(user) && current_user.friend_requests.none?(user) && current_user != user %>
+      <%= link_to 'Add as friend', user_friendships_path(current_user, :friendship => {:friend_id => user.id}), method: :post, class: 'btn-1 green_btn' %>
+    <% elsif (current_user.pending_friends.include?(user) || current_user.friend_requests.include?(user)) %>
+      <p>'Waiting for confirmation.'</p>
+    <% elsif current_user != user %>
+      <p>'You are already friends.'</p>
+    <% end %>
+  </div>
+  <div class="Rtable-cell">&nbsp;</div>
+  <% end %>
```
Comment (tirthajyoti-ghosh) [inline app/views/users/_user.html.erb]: Please move this logic to helpers.

File: app/views/users/index.html.erb [hunk 5]
```diff
@@ -1,5 +1,32 @@
...
+<section class="friend-requests-section">
+  <ul class="users-list">
+    <h2>Received Requests</h2>
+    <% @friend_requests.each do |user| %>
+    <li class="my-4">
+      <% if !current_user.friend?(user) %>
+        <%= user.name %> want to be your friend.
+        <%= link_to 'Accept', user_friendship_path(@friend_requests, user), method: :put, style: "color: #008000" %>
+        <%= link_to 'Reject', user_friendship_path(@friend_requests, user), method: :delete, style: "color: #FF0000" %>
+      <% end %>
```
Comment (tirthajyoti-ghosh) [inline app/views/users/index.html.erb]: Please move this logic to helpers.

File: app/views/users/index.html.erb [hunk 6]
```diff
@@ -1,5 +1,32 @@
...
+  </ul>
+  
   <ul class="users-list">
-    <%= render @users %>
+    <h2>Sent Requests</h2>
+    <% @pending_requests.each do |user| %>
+    <li class="my-4">
+      <% if !current_user.friend?(user) %>
+      Waiting friend confirmation from <%= user.name %>
+      <% end %>
```
Comment (tirthajyoti-ghosh) [inline app/views/users/index.html.erb]: Please move this logic to helpers.

File: app/views/users/show.html.erb [hunk 7]
```diff
@@ -1,9 +1,24 @@
 <section class="user-section">
   <h2><%= "Name: #{@user.name}" %></h2>
+
+  <% if current_user.friends.none?(@user) && current_user.pending_friends.none?(@user) && current_user != @user %>
+    <%= link_to 'Add as friend', user_friendships_path(current_user, :friendship => {:friend_id => @user.id}), method: :post, class: 'btn-1 green_btn' %>
+  <% elsif current_user.friends.none?(@user) && current_user.pending_friends.include?(@user) %>
+    <p>'Waiting for confirmation.'</p>
+  <% elsif current_user != @user %>
+    <p>'You are already friends.'</p>
+  <% end %>
```
Comment (tirthajyoti-ghosh) [inline app/views/users/show.html.erb]: Please move this logic to helpers.

**Review by tirthajyoti-ghosh — CHANGES_REQUESTED** [review]
# Changes required 🛑 

## Impressive progress so far 💪! Thanks for making all the changes 👏.

### However, we need to work on some changes before we can merge.

There is still some logic left in your views after the previous review. I have marked them. 
  - [ ] Please go over and move them to helpers.

That's it!

Your project looks absolutely fantastic. Just make these changes and apply for another code review.

Happy coding 👨‍💻 

*Need help? Shoot me a message on Slack - [@Ghosh](https://app.slack.com/client/T47CT8XPG/DRDM6LAQN/user_profile/URB1MHXK6)*
<hr>

**Tirthajyoti Ghosh**

[GitHub](https://github.com/tirthajyoti-ghosh) | [LinkedIn](https://www.linkedin.com/in/tirthajyoti-ghosh/) | [Twitter](https://twitter.com/terrific_ghosh)

**Review by ghost — CHANGES_REQUESTED** [review]
# Status: Changes Required 

## Hello team 😃,

**Great job on your project 💪. However, the changes requested by the previous TSE haven't been implemented yet. The only change made was in the `app/views/user/index.html.erb` file. Please work on the rest as well.**

### Happy coding 🥂.

**PS: You can reach me on slack (Rose Wanjohi).**

**Review by ghost — CHANGES_REQUESTED** [review]
# Status: Changes Required 

**You are almost there team 💪. Here is the list of files that still have logic in them.**

- `app/views/layouts/application.html.erb`.
- `app/views/users/_user.html.erb `.

I also noticed your RSpec tests are not passing anymore.

rspec

**Review by acushlakoncept — APPROVED** [review]
# Status: PROJECT APPROVED  🎉 🚀

I see both of you have made lots of changes to meet all requirements, I am sure you will be proud of yourselves at the end of this project. 
#### Congratulations you've both done well on this milestone 👏 👏 👏 

I will go ahead and mark this as complete, feel free to merge when you are ready. All the best with the next milestone.

### Cheers and Happy coding! 💻 👏

(Feel free to contact me on slack(@acushla), if you have any additional question)
Uduak Essien
[Twitter](https://twitter.com/acushlakoncept) | [LinkedIn](https://www.linkedin.com/in/acushlakoncept/)

**Pull request merged**